contracts/hydro/src/contract.rs

```diff
@@ -715,71 +736,89 @@ fn vote(
                         vote.prop_id,
                     ),
-                    &vote.prop_id,
+                );
+
+                remove_validator_shares_from_proposal(
+                    deps.storage,
+                    round_id,
+                    vote.prop_id,
+                    vote.time_weighted_shares.0,
+                    vote.time_weighted_shares.1,
                 )?;
-            }
 
-            // Delete vote
-            VOTE_MAP.remove(
-                deps.storage,
-                ((round_id, tranche_id), info.sender.clone(), lock_id),
-            );
+                // save the new power into the proposal
+                let total_power =
+                    get_total_power_for_proposal(deps.as_ref().storage, vote.prop_id)?;
+                proposal.power = total_power.to_uint_ceil(); // TODO: decide whether we need to round or represent as decimals
 
-            response = response.add_attribute(
-                format!("lock_id_{}_old_proposal_id", lock_id),
-                vote.prop_id.to_string(),
-            );
-        }
+                // Save the proposal
+                PROPOSAL_MAP.save(
+                    deps.storage,
+                    (round_id, tranche_id, vote.prop_id),
+                    &proposal,
+                )?;
+
+                // Add proposal's new power in PROPS_BY_SCORE
+                if proposal.power > Uint128::zero() {
+                    PROPS_BY_SCORE.save(
+                        deps.storage,
+                        (
+                            (round_id, proposal.tranche_id),
+                            proposal.power.into(),
+                            vote.prop_id,
+                        ),
+                        &vote.prop_id,
+                    )?;
+                }
+
+                // Delete vote
+                VOTE_MAP.remove(
+                    deps.storage,
+                    ((round_id, tranche_id), info.sender.clone(), lock_id),
+                );
+
+                response = response.add_attribute(
+                    format!("lock_id_{}_old_proposal_id", lock_id),
+                    vote.prop_id.to_string(),
+                );
+            }
+            None => {
+                // If user didn't yet vote with the given lock in the given round and tranche, check if they
+                // voted in previous rounds for some proposal that spans multiple rounds.
+                let voting_allowed_round = VOTING_ALLOWED_ROUND
+                    .may_load(deps.storage, (tranche_id, info.sender.clone(), lock_id))?;
+
+                if let Some(voting_allowed_round) = voting_allowed_round {
+                    if voting_allowed_round > round_id {
+                        return Err(ContractError::Std(
+                        StdError::generic_err(format!(
+                            "Not allowed to vote with lock_id {} on new proposal. Voted already on proposal still in bidding phase ending in round {}.",
```

Comment:
```suggestion
                            "Not allowed to vote with lock_id {} in tranche {}: Voted in round {} on proposal {} with duration {}; cannot vote again with this lock_id until round {}",
```
error should be a bit more descriptive imo

Reply:
Currently we can't extract the 'long lasting' proposal that user voted on. If you still think we should change this message, then we should store the round id in which voting occurred in the `VOTING_ALLOWED_ROUND`. WDYT, is it worth it?

Reply:
Let's just leave the proposal and round out and just say when they can vote again then